### PR TITLE
feat: refactor rule-maker

### PR DESCRIPTION
**Pull request by art920**
# Description

What is the overall goal of your PR? Which problem does it solve? Please also include relevant motivation and context.
List any dependencies that are required for this change.

Fixes #(issue number)

# Migrations required

yes: please describe the migration
no: please delete the whole paragraph

# Verification

Please describe the test cases you used to verify your code. Did you check the change in your environment?

# Checklist

- [ ] My code follows the style guidelines of the project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/main/java/com/hlag/rulemaker/expression/CustomMathExpression.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package com.hlag.rulemaker.expression;
+
+import io.github.jamsesso.jsonlogic.evaluator.JsonLogicEvaluationException;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.function.BinaryOperator;
+
+public class CustomMathExpression extends NumericExpression {
```
Comment (poplawskidaniel) [inline src/main/java/com/hlag/rulemaker/expression/CustomMathExpression.java:8]: BigDecimalMathExpression ?

File: src/main/java/com/hlag/rulemaker/RuleMaker.java [hunk 2]
```diff
@@ -1,631 +1,614 @@
...
+   * @param value The value to check.
+   * @param set   The array to check against or string to check for substring.
+   * @return A new RuleMaker representing the in operation.
+   */
+  public static RuleMaker in(RuleMaker value, RuleMaker set) {
+    return template("in", value, set);
+  }
+
+  /**
+   * Creates a RuleMaker representing a cat operation.
```
Comment (poplawskidaniel) [inline src/main/java/com/hlag/rulemaker/RuleMaker.java:439]: cat operation?

File: src/main/java/com/hlag/rulemaker/RuleMaker.java [hunk 3]
```diff
@@ -1,631 +1,614 @@
...
+   * @param start  The start index of the substring.
+   * @param length The length of the substring.
+   * @return A new RuleMaker representing a substring operation.
+   */
+  public static RuleMaker substr(RuleMaker value, RuleMaker start, RuleMaker length) {
+    return template("substr", value, start, length);
+  }
+
+  /**
+   * Creates a RuleMaker representing a log operation.
```
Comment (poplawskidaniel) [inline src/main/java/com/hlag/rulemaker/RuleMaker.java:461]: log operation? Please make this more precise is it a logging or logarithm etc?

File: src/test/java/com/hlag/rulemaker/expression/DateDiffExpressionUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    //When
+    //Then
+    assertThatThrownBy(() -> DateDiffExpression.INSTANCE.evaluate(arguments, null))
+      .isInstanceOf(IllegalArgumentException.class)
+      .hasMessage("Unexpected value 'Incorrect'");
+  }
+
+  private static Stream<Arguments> provideCorrectTestData() {
+    return Stream.of(
+      Arguments.of("2024-04-19", "2024-04-20", DayType.CALENDAR_DAYS, 1L),
```
Comment (poplawskidaniel) [inline src/test/java/com/hlag/rulemaker/expression/DateDiffExpressionUnitTest.java:73]: Please add more complicated examples like between years: "2021-12-12" "2022-01-18" etc.

File: src/test/java/com/hlag/rulemaker/expression/ClampExpressionUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      Arguments.of(List.of(new BigDecimal("2.0"), new BigDecimal("1.0"), new BigDecimal("3.0")), new BigDecimal("2.0")),
+      Arguments.of(List.of(new BigDecimal("2.0"), new BigDecimal("3.0"), new BigDecimal("4.0")), new BigDecimal("3.0")),
+      Arguments.of(List.of(new BigDecimal("4.0"), new BigDecimal("1.0"), new BigDecimal("3.0")), new BigDecimal("3.0")),
+      Arguments.of(List.of(new BigDecimal("2.0"), new BigDecimal("2.0"), new BigDecimal("2.0")), new BigDecimal("2.0"))
+    );
+  }
+
+  private static Stream<Arguments> incorrectNumberOfArguments() {
+    return Stream.of(
+      Arguments.of(List.of()),
```
Comment (poplawskidaniel) [inline src/test/java/com/hlag/rulemaker/expression/ClampExpressionUnitTest.java:51]: What about cases that value is not between min and max ?

File: src/main/java/com/hlag/rulemaker/RuleMakerJsonLogic.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import com.hlag.rulemaker.expression.CustomNumericComparisonExpression;
+import com.hlag.rulemaker.expression.DateDiffExpression;
+import io.github.jamsesso.jsonlogic.JsonLogic;
+import io.github.jamsesso.jsonlogic.JsonLogicException;
+
+public class RuleMakerJsonLogic {
+
+    private final JsonLogic jsonLogic;
+
+    public RuleMakerJsonLogic() {
```
Comment (poplawskidaniel) [inline src/main/java/com/hlag/rulemaker/RuleMakerJsonLogic.java:14]: at least tests that check that all custom operation are added should be added.

**Review by poplawskidaniel — CHANGES_REQUESTED** [review]

File: src/main/java/com/hlag/rulemaker/expression/CustomMathExpression.java (hunk 1 above)
Comment (art920) [reply src/main/java/com/hlag/rulemaker/expression/CustomMathExpression.java:8]: done

File: src/main/java/com/hlag/rulemaker/RuleMaker.java (hunk 2 above)
Comment (art920) [reply src/main/java/com/hlag/rulemaker/RuleMaker.java:439]: added description

File: src/main/java/com/hlag/rulemaker/expression/ClampExpression.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  @SuppressWarnings({"squid:S1659"})
+  @Override
+  public Object evaluate(List arguments, Object data) throws JsonLogicEvaluationException {
+    if (arguments.size() != 3) {
+      throw new JsonLogicEvaluationException("clamp expects exactly 3 arguments: value, min, max");
+    }
+
+    BigDecimal[] values = parseArguments(arguments);
```
[the remark this answers is not in the document]
Comment (art920) [reply src/main/java/com/hlag/rulemaker/expression/ClampExpression.java:29]: we shouldn't do that, the value can be lower than min and higher than max, in this cases min and max are applied

File: src/test/java/com/hlag/rulemaker/expression/DateDiffExpressionUnitTest.java (hunk 4 above)
Comment (art920) [reply src/test/java/com/hlag/rulemaker/expression/DateDiffExpressionUnitTest.java:73]: done

File: src/main/java/com/hlag/rulemaker/RuleMakerJsonLogic.java (hunk 6 above)
Comment (art920) [reply src/main/java/com/hlag/rulemaker/RuleMakerJsonLogic.java:14]: tested in `RuleMakerUnitTest`

**Review by poplawskidaniel — APPROVED** [review]

**Pull request merged**